File: common/src/main/java/cc/turtl/chiselmon/command/DashCommand.java
package cc.turtl.chiselmon.command;

import cc.turtl.chiselmon.system.tracker.TrackerManager;
import cc.turtl.chiselmon.system.tracker.TrackerSession;
import cc.turtl.chiselmon.util.MessageUtils;
import cc.turtl.chiselmon.util.format.ColorUtils;
import cc.turtl.chiselmon.util.format.ComponentUtils;
import cc.turtl.chiselmon.util.format.StringFormats;
import com.mojang.brigadier.Command;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.context.CommandContext;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.network.chat.Component;

import java.io.IOException;

public class DashCommand implements ChiselmonCommand {

    @Override
    public String getName() {
        return "dash";
    }

    @Override
    public String getDescription() {
        return "View detailed spawning stats in the Chiselmon Dash.";
    }

    @Override
    public <S> LiteralArgumentBuilder<S> build() {
        return LiteralArgumentBuilder.<S>literal(getName())
                .executes(this::showHelp)
                .then(LiteralArgumentBuilder.<S>literal("status")
                        .executes(this::executeStatus))
                .then(LiteralArgumentBuilder.<S>literal("open")
                        .executes(this::executeOpen))
                .then(LiteralArgumentBuilder.<S>literal("close")
                        .executes(this::executeClose));
    }

    private <S> int showHelp(CommandContext<S> context) {
        LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) return 0;

        String root = context.getNodes().getFirst().getNode().getName();

        MessageUtils.sendEmptyLine(player);
        MessageUtils.sendSuccess(player, "Chiselmon Dash - Commands");
        MessageUtils.sendPrefixed(player, "  /" + root + " dash status");
        MessageUtils.sendPrefixed(player, "  /" + root + " dash open");
        MessageUtils.sendPrefixed(player, "  /" + root + " dash close");
        return Command.SINGLE_SUCCESS;
    }

    private <S> int executeStatus(CommandContext<S> context) {
        LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) return 0;

        TrackerSession session = requireSession(player);
        if (session == null) return Command.SINGLE_SUCCESS;

        if (!session.isDashboardRunning()) {
            MessageUtils.sendWarning(player, "Dashboard server is not running.");
            return Command.SINGLE_SUCCESS;
        }

        String url = "http://localhost:" + session.getDashboardPort() + "/";

        String uptime = StringFormats.formatDurationMs(session.dashboardUptime());

        MessageUtils.sendPrefixed(player, Component.literal("Dashboard server is running at ")
                .withColor(ColorUtils.GREEN.getRGB())
                .append(ComponentUtils.clickableUrl(url))
                .append(Component.literal(" (Uptime: " + uptime + ")")));
        return Command.SINGLE_SUCCESS;
    }

    private <S> int executeOpen(CommandContext<S> context) {
        LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) return 0;

        TrackerSession session = requireSession(player);
        if (session == null) return Command.SINGLE_SUCCESS;

        String url = "http://localhost:" + session.getDashboardPort() + "/";

        if (session.isDashboardRunning()) {
            MessageUtils.sendPrefixed(player, Component.literal("Dashboard is already running at ")
                    .withColor(ColorUtils.YELLOW.getRGB())
                    .append(ComponentUtils.clickableUrl(url)));
            return Command.SINGLE_SUCCESS;
        }

        try {
            session.startDashboard();
            MessageUtils.sendPrefixed(player, Component.literal("Dashboard opened at ")
                    .withColor(ColorUtils.GREEN.getRGB())
                    .append(ComponentUtils.clickableUrl(url)));
        } catch (IOException e) {
            MessageUtils.sendError(player, context, e);
        }

        return Command.SINGLE_SUCCESS;
    }

    private <S> int executeClose(CommandContext<S> context) {
        LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) return 0;

        TrackerSession session = requireSession(player);
        if (session == null) return Command.SINGLE_SUCCESS;

        if (!session.isDashboardRunning()) {
            MessageUtils.sendWarning(player, "Dashboard server is not running.");
            return Command.SINGLE_SUCCESS;
        }

        session.stopDashboard();
        MessageUtils.sendSuccess(player, "Dashboard server closed.");
        return Command.SINGLE_SUCCESS;
    }

    /**
     * Validates a session exists, sending a warning if not. Returns null if invalid.
     */
    private TrackerSession requireSession(LocalPlayer player) {
        try {
            return TrackerManager.getInstance().getTracker();
        } catch (IllegalStateException e) {
            MessageUtils.sendWarning(player, "No active tracker session.");
            return null;
        }
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/command/DashCommand.kt
package cc.turtl.chiselmon.client.command

import cc.turtl.chiselmon.client.util.*
import cc.turtl.chiselmon.core.util.format.clickableUrl
import cc.turtl.chiselmon.client.system.tracker.TrackerSession
import cc.turtl.chiselmon.client.system.tracker.TrackerSession
import cc.turtl.turtlshell.api.core.command.TurtlShellCommand
import cc.turtl.turtlshell.api.core.format.ColorLib
import cc.turtl.turtlshell.api.core.format.formatDuration
import com.mojang.brigadier.Command
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import net.minecraft.client.Minecraft
import net.minecraft.client.player.LocalPlayer
import net.minecraft.commands.CommandSourceStack
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent

class DashCommand : TurtlShellCommand {

    override val name = "dash"
    override val description: MutableComponent = Component.literal("View detailed spawning stats in the Chiselmon Dash")

    override fun build(): LiteralArgumentBuilder<CommandSourceStack> =
        LiteralArgumentBuilder.literal<CommandSourceStack>(name)
            .executes { ctx ->
                val player = Minecraft.getInstance().player ?: return@executes 0
                val root = ctx.nodes.first().node.name

                sendEmptyLine(player)
                sendSuccess(player, "Chiselmon Dash - Commands")
                sendPrefixed(player, "  /$root dash status")
                sendPrefixed(player, "  /$root dash open")
                sendPrefixed(player, "  /$root dash close")
                Command.SINGLE_SUCCESS
            }
            .then(
                LiteralArgumentBuilder.literal<CommandSourceStack>("status")
                    .executes { executeStatus() }
            )
            .then(
                LiteralArgumentBuilder.literal<CommandSourceStack>("open")
                    .executes { executeOpen() }
            )
            .then(
                LiteralArgumentBuilder.literal<CommandSourceStack>("close")
                    .executes { executeClose() }
            )

    private fun executeStatus(): Int {
        val player = Minecraft.getInstance().player ?: return 0
        val session = requireSession(player) ?: return Command.SINGLE_SUCCESS

        if (!session.isDashboardRunning) {
            sendWarning(player, "Dashboard server is not running.")
            return Command.SINGLE_SUCCESS
        }

        val url = "http://localhost:${session.dashboardPort}/"
        val uptime = formatDuration(session.dashboardUptime())

        sendPrefixed(
            player, Component.literal("Dashboard server is running at ")
                .withColor(ColorLib.GREEN.rgb)
                .append(clickableUrl(url))
                .append(Component.literal(" (Uptime: $uptime)"))
        )
        return Command.SINGLE_SUCCESS
    }

    private fun executeOpen(): Int {
        val player = Minecraft.getInstance().player ?: return 0
        val session = requireSession(player) ?: return Command.SINGLE_SUCCESS

        val url = "http://localhost:${session.dashboardPort}/"

        if (session.isDashboardRunning) {
            sendPrefixed(
                player, Component.literal("Dashboard is already running at ")
                    .withColor(ColorLib.YELLOW.rgb)
                    .append(clickableUrl(url))
            )
            return Command.SINGLE_SUCCESS
        }

        try {
            session.startDashboard()
            sendPrefixed(
                player, Component.literal("Dashboard server opened at ")
                    .withColor(ColorLib.GREEN.rgb)
                    .append(clickableUrl(url))
            )
        } catch (e: Exception) {
            sendError(player, e)
        }

        return Command.SINGLE_SUCCESS
    }

    private fun executeClose(): Int {
        val player = Minecraft.getInstance().player ?: return 0
        val session = requireSession(player) ?: return Command.SINGLE_SUCCESS

        if (!session.isDashboardRunning) {
            sendWarning(player, "Dashboard server is not running.")
            return Command.SINGLE_SUCCESS
        }

        session.stopDashboard()
        sendSuccess(player, "Dashboard server closed.")
        return Command.SINGLE_SUCCESS
    }

    /**
     * Validates a session exists, sending a warning if not. Returns null if invalid.
     */
    private fun requireSession(player: LocalPlayer): TrackerSession? {
        return try {
            TrackerSession.current
        } catch (e: IllegalStateException) {
            sendWarning(player, "No active tracker session.")
            null
        }
    }
}